### PR TITLE
Fix I2C signalStop in readLastByte for Microbit

### PR DESCRIPTION
**Pull request by torntrousers**
Fix for https://github.com/tinygo-org/tinygo/issues/343

### CODE REVIEW

File: src/machine/machine_nrf.go [hunk 1]
```diff
@@ -234,8 +234,8 @@ func (i2c I2C) readByte() byte {
 func (i2c I2C) readLastByte() byte {
 	for i2c.Bus.EVENTS_RXDREADY == 0 {
 	}
-	i2c.Bus.EVENTS_RXDREADY = 0
 	i2c.signalStop() // signal 'stop' now, so it is sent when reading RXD
+	i2c.Bus.EVENTS_RXDREADY = 0
```
Comment (deadprogram) [inline src/machine/machine_nrf.go]: What if you remove this line entirely from here?

Comment (torntrousers) [reply src/machine/machine_nrf.go]: No trying that then it doesn't work at all.

**Comment by torntrousers** [comment]
Closing as it doesn't work

**Comment by torntrousers** [comment]
Reopening with new code based on the suggestion from @deadprogram in #343

**Comment by torntrousers** [comment]
I've tested this with the BMP180 and SHT3x examples and the u-blox GPS and it seems to work fine.

File: src/machine/machine_nrf.go [hunk 2]
```diff
@@ -185,21 +185,19 @@ func (i2c I2C) Tx(addr uint16, w, r []byte) error {
...
 		for i := range r {        // read each char
 			if i+1 == len(r) {
-				// The 'stop' signal must be sent before reading back the last
-				// byte, so that it will be sent by the I2C peripheral right
-				// after the last byte has been read.
-				r[i] = i2c.readLastByte()
-			} else {
-				r[i] = i2c.readByte()
+				// To trigger stop task when last byte is received, set before resume task.
+				i2c.Bus.SHORTS = 2
```
Comment (deadprogram) [inline src/machine/machine_nrf.go]: This should probably be `i2c.Bus.SHORTS = nrf.TWI_SHORTS_BB_STOP`

File: src/machine/machine_nrf.go [hunk 3]
```diff
@@ -185,21 +185,19 @@ func (i2c I2C) Tx(addr uint16, w, r []byte) error {
...
 			}
+			i2c.Bus.TASKS_RESUME = 1 // re-start transmission for reading
+			r[i] = i2c.readByte()
 		}
-	} else {
-		// Nothing to read back. Stop the transmission.
-		i2c.signalStop()
 	}
+	i2c.signalStop()
+	i2c.Bus.SHORTS = 0
```
Comment (deadprogram) [inline src/machine/machine_nrf.go]: I think this should be `i2c.Bus.SHORTS &^= (nrf.TWI_SHORTS_BB_STOP << nrf.TWI_SHORTS_BB_STOP_Pos)`

File: src/machine/machine_nrf.go [hunk 4]
```diff
@@ -185,21 +185,21 @@ func (i2c I2C) Tx(addr uint16, w, r []byte) error {
 		}
 	}
 	if len(r) != 0 {
+		// To trigger suspend task when a byte is received
+		i2c.Bus.SHORTS &^= (nrf.TWI_SHORTS_BB_STOP << nrf.TWI_SHORTS_BB_STOP_Pos)
```
Comment (deadprogram) [inline src/machine/machine_nrf.go]: I'm sorry that I made a mistake before. The correct syntax here would be better stated `i2c.Bus.SHORTS &^= nrf.TWI_SHORTS_BB_STOP` like you have below.

Comment (torntrousers) [reply src/machine/machine_nrf.go]: I'm not sure I understand - we want i2c.Bus.SHORTS to be set to 1 here, don't we? &^= nrf.TWI_SHORTS_BB_STOP isn't going to do that is it?

Comment (deadprogram) [reply src/machine/machine_nrf.go]: In that case you want `i2c.Bus.SHORTS = nrf.TWI_SHORTS_BB_STOP` as the `&^=` operator turns the bit OFF.

Comment (torntrousers) [reply src/machine/machine_nrf.go]: How about like this with TWI_SHORTS_BB_SUSPEND?

File: src/machine/machine_nrf.go [hunk 5]
```diff
@@ -185,21 +185,21 @@ func (i2c I2C) Tx(addr uint16, w, r []byte) error {
 		}
 	}
 	if len(r) != 0 {
+		// To trigger suspend task when a byte is received
+		i2c.Bus.SHORTS = nrf.TWI_SHORTS_BB_SUSPEND
 		i2c.Bus.TASKS_STARTRX = 1 // re-start transmission for reading
 		for i := range r {        // read each char
+			// i2c.Bus.TASKS_RESUME = 1 // re-start transmission for reading
```
Comment (deadprogram) [inline src/machine/machine_nrf.go]: Is this supposed to be commented out? If so it should be removed instead.

File: src/machine/machine_nrf.go (hunk 4 above)
Comment (deadprogram) [reply src/machine/machine_nrf.go]: Now I am confused, sorry. What is it exactly you want to set or turn off here?

Comment (torntrousers) [reply src/machine/machine_nrf.go]: I think the idea is to:
- at the start of a read set SHORTS to 1, which is TWI_SHORTS_BB_SUSPEND
- before the last byte of the read is read set SHORTS to 2, which is TWI_SHORTS_BB_STOP
- when the read is finished set SHORTS to 0, which is &^= nrf.TWI_SHORTS_BB_STOP? Or perhaps that should be TWI_SHORTS_BB_STOP_Disabled?

Comment (deadprogram) [reply src/machine/machine_nrf.go]: Seems like that would be best expressed as:

```
i2c.Bus.SHORTS = sam.TWI_SHORTS_BB_SUSPEND
```

```
i2c.Bus.SHORTS = sam.TWI_SHORTS_BB_STOP
```

```
i2c.Bus.SHORTS = sam.TWI_SHORTS_BB_SUSPEND_Disabled
```

**Comment by deadprogram** [comment]
@torntrousers thank you so very much for all the hard work on getting this fix in. Now merging, thanks again.

**Comment by deadprogram** [comment]
I meant to merge into dev not master, so I had to revert, and cherry pick into dev. Anyhow, has been corrected.

**Pull request merged**